File: lib/comet-html/generator.rb
require 'comet-html/parser-class'
require 'comet-html/parser-repeater'
require 'comet-html/parser-slot'
require 'comet-html/utils'
require 'comet-html/header-generator'
require 'comet-html/source-generator'
require 'nokogiri'
require 'pathname'

module Comet
  class LoadError < StandardError
    attr_reader :message
    def initialize
      @message = "Load error: #{message}"
    end
  end

  class ParseError < StandardError
    attr_reader :el, :message

    def initialize el, message
      @el = el
      @message = message
    end
    
    def line
      el.line
    end
  end
  
  class HtmlParseError < StandardError
    attr_reader :message, :line
    
    def initialize error
      @message = error.message
      @line    = error.line
    end
  end

  class Generator
    def initialize output, input, file
      @output   = output
      @input    = input
      @filepath = file
    end

    def set_config config
      @config = config
    end
    
    def relative_filepath
      @filepath[@input.length..-1]
    end

    def compiled_header_path
      @output + relative_filepath.delete_suffix(File.extname relative_filepath) + ".hpp"
    end

    def compiled_source_path
      @output + relative_filepath.delete_suffix(File.extname relative_filepath) + ".cpp"
    end
  
    def append_classes_predeclaration
      result = ""
      Context.classes.each do |object|
        next if object.should_skip?
        result += "  class #{object.typename};\n"
      end
      result
    end
    
    def load_includes head
      head.css("include").each do |include_attribute|
        unless include_attribute["require"].nil?
          type = include_attribute["require"].to_s
          tag  = if include_attribute["tag-name"].nil?
            parts = type.split("<")
            parts[0].gsub! "::", "_"
            parts.join("<").dasherize
          else
            include_attribute["tag-name"].to_s
          end
          Context.element_types[tag] = type
        end
      end
    end

    def prepare_includes head
      result = ""
      included_paths = []
      head.css("include").each do |include_attribute|
        included_paths << include_attribute["src"].to_s
      end
      included_paths.uniq.each do |include_path|
        result += "# include \"#{include_path}\"\n"
      end
      @config["elements"].each do |element_data|
        if Context.referenced_types.include? element_data["require"]
          result += "# include \"#{element_data["include"]}\"\n"
        end
      end if !@config.nil? && !@config["elements"].nil?
      result
    end

    def html5_tags
      ["article", "aside", "details", "figcaption", "figure", "footer", "header", "main", "mark", "nav", "section", "summary", "time"]
    end
    
    def whitelisted_tags
      ["svg"] + (if @config.nil? then [] else @config["whitelisted-tags"] || [] end)
    end
    
    def internal_tags
      ["include", "attribute", "slot"]
    end

    def is_error_fatal? error
      matches = error.message.match(/^[0-9]+:[0-9]+: ERROR: Tag (.+) invalid$/)
      unless matches.nil?
        tagName = matches[1]
        return false if (internal_tags + html5_tags + whitelisted_tags).include? tagName
        return (if Context.element_types[tagName].nil? then true else false end)
      end
      return false if error.message.end_with?("ERROR: htmlParseEntityRef: no name") ||
                      error.message.end_with?("ERROR: Expected a doctype token$")
      true
    end

    def generate
      src = File.read @filepath
      src.gsub! /<template/i,  "<html"
      src.gsub! /<\/template/i, "</html>"

      document = Nokogiri::HTML.parse(src) do |config| config.strict end
      
      html = document.xpath("html").first
      head = document.xpath("//head")
      body = document.xpath("//body")
      raise LoadError.new "could not load html template `#{@filepath}`" if html.nil?
      raise LoadError.new "missing body element in `#{@filepath}`" if body.first.nil?

      Context.reset
      Context.filename = @filepath
      if !@config.nil? && !@config["elements"].nil?
        Context.load_global_element_types @config["elements"]
      end
      
      load_includes head
      
      fatal_error = nil
      document.errors.each do |error|
        if is_error_fatal? error
          puts "[comet-html] #{@filepath}:#{error.line}:#{error.column}: #{error.message}"
          fatal_error ||= error
          error_count += 1
        end
      end
      raise HtmlParseError.new fatal_error unless fatal_error.nil?

      main_element = Comet::Class.new body
      main_element.typename = if html["classname"].nil?
        File.basename(@filepath, ".html").to_s.camelize
      else
        html["classname"]
      end
      main_element.superclass = html["extends"] unless html["extends"].nil?
      main_element.probe

      main_element.inline_code = ""
      head.css("script").each do |script|
        main_element.inline_code += script.text.strip + "\n"
      end

      head.css("attribute").each do |attribute|
        main_element.refs << (CppReference.new attribute, attribute["type"], attribute["name"], attribute["value"])
      end

      define_name = "__COMET_CUSTOM_ELEMENT_#{main_element.typename.upcase}__"
      header_code = ""
      source_code = ""
      Context.classes.reverse.each do |object|
        next if object.should_skip?
        header_code += HeaderGenerator.new(object).generate + "\n"
        source_code += SourceGenerator.new(object).generate + "\n"
      end

      includes_str = prepare_includes head
      
      header = <<HEADER
#ifndef  #{define_name}
# define #{define_name}

# include <comet/bindable.hpp>
# include <comet/repeater.hpp>
# include <comet/slot_element.hpp>
# include <comet/comment_element.hpp>
# include <comet/custom_element.hpp>
# include <comet/signal.hpp>
#{includes_str}

namespace HtmlTemplate
{
#{append_classes_predeclaration}
#{header_code}
}

#endif
HEADER

      source = <<SOURCE
#include <comet/lexical_cast.hpp>
#include "#{compiled_header_path}"

#{source_code}
SOURCE

      { header: header, source: source }
    end
  end
end

File: comet-html/generator.cpp
#include "generator.hpp"
#include "probe.hpp"
#include "header_generator.hpp"
#include <set>
#include <crails/utils/semantics.hpp>

using namespace std;

void Generator::prepare()
{
  auto html = document.select_node("template");
  auto head = document.select_node("//head");
  auto body = document.select_node("//body");

  if (html.node().type() == pugi::node_null) throw LoadError("could not find <template> element");
  if (body.node().type() == pugi::node_null) throw LoadError("missing <body> element");
  this->head = head.node();
  this->body = body.node();
  this->head.select_nodes("//attribute");

  if (config.if_contains("elements"))
    Context::global.load_global_element_types(config.at("elements").as_array());
  if (head.node().type() != pugi::node_null)
    load_includes();
  root = Context::make<Class>(body.node());
  initialize_classname(html.node());
  initialize_superclass(html.node());
  initialize_attributes();
  Probe::probe(*root);
}

void Generator::set_inline_code(const string& inline_code)
{
  root->inline_code = inline_code;
}

string Generator::generate_header()
{
  stringstream stream;
  string guard_name = "__COMET_CUSTOM_ELEMENT_" + root->get_typename() + "__";
  auto classes = generated_classes();

  stream << "#ifndef " << guard_name << endl
          << "# define " << guard_name << endl
          << "# include <comet/bindable.hpp>" << endl
          << "# include <comet/repeater.hpp>" << endl
          << "# include <comet/slot_element.hpp>" << endl
          << "# include <comet/comment_element.hpp>" << endl
          << "# include <comet/custom_element.hpp>" << endl
          << "# include <comet/signal.hpp>" << endl
          << prepare_includes() << endl
          << "namespace HtmlTemplate" << endl
          << '{' << endl;
  for (const auto& class_ : classes)
    stream << "  class " << class_->get_typename() << ';' << endl;
  for (auto it = classes.rbegin() ; it != classes.rend() ; ++it)
    stream << HeaderGenerator(**it).generate() << endl;
  stream << '}' << endl << "#endif";
  return stream.str();
}

string Generator::generate_source()
{
  stringstream stream;

  stream << "#include <comet/lexical_cast.hpp>" << endl
          << "#include \"" << compiled_header_path() << '"' << endl << endl;
  for (const auto& class_ : generated_classes())
    stream << SourceGenerator(*class_).generate() << endl;
  return stream.str();
}

vector<shared_ptr<Class>> Generator::generated_classes()
{
  vector<shared_ptr<Class>> classes;

  for (const auto& class_ : Context::global.classes)
  {
    if (!class_->should_skip()) classes.push_back(class_);
  }
  return classes;
}

void Generator::initialize_classname(const pugi::xml_node& node)
{
  pugi::xml_attribute attribute = node.attribute("classname");

  if (attribute.empty())
  {
    auto filename = Context::global.file_path.filename();
    auto classname = filename.string().substr(0, filename.string().length() - filename.extension().string().length());
    root->type_name = Crails::camelize(classname);
  }
  else
    root->type_name = attribute.value();
}

void Generator::initialize_superclass(const pugi::xml_node& node)
{
  pugi::xml_attribute attribute = node.attribute("extends");

  if (!attribute.empty())
    root->super_class = attribute.value();
}

void Generator::initialize_attributes()
{
  for (pugi::xpath_node xpath : head.select_nodes("//attribute"))
  {
    pugi::xml_node attribute = xpath.node();
    pugi::xml_attribute type = attribute.attribute("type");
    pugi::xml_attribute name = attribute.attribute("name");
    pugi::xml_attribute value = attribute.attribute("value");

    root->references.push_back(make_shared<CppReference>(attribute, type.value(), name.value(), value.value()));
  }
}

void Generator::load_includes()
{
  pugi::xpath_node_set includes = head.select_nodes("include");

  for (const pugi::xpath_node& xpath_node : includes)
  {
    pugi::xml_node node = xpath_node.node();
    pugi::xml_attribute require = node.attribute("require");
    pugi::xml_attribute tag_attribute = node.attribute("tag-name");

    if (!require.empty())
    {
      string type = require.value();
      string tag_name;

      if (tag_attribute.empty())
        tag_name = Crails::dasherize(type); // TODO: replace "::" with "_", remove anything after '<'
      else
        tag_name = tag_attribute.value();
      Context::global.element_types.emplace(tag_name, type);
    }
  }
}

string Generator::prepare_includes()
{
  stringstream stream;
  set<string> paths;
  const auto& referenced_types = Context::global.referenced_types;

  // Explicits includes in <head>
  for (const pugi::xpath_node& xpath_node : head.select_nodes("include"))
    paths.insert(xpath_node.node().attribute("src").value());
  // Implicits includes from JSON config
  if (config.if_contains("elements"))
  {
    for (const boost::json::value& element_value : config.at("elements").as_array())
    {
      const boost::json::object element = element_value.as_object();
      string require = boost::json::value_to<string>(element.at("require"));

      if (find(referenced_types.begin(), referenced_types.end(), require) != referenced_types.end())
        paths.insert(boost::json::value_to<string>(element.at("include")));
    }
  }
  for (const string& path : paths)
    stream << "# include \"" + path + "\"" << endl;
  return stream.str();
}

static string replace_extension(const string& path, const string& extension) { return path.substr(0, path.find_last_of(".")) + '.' + extension; }

string Generator::relative_filepath() const
{
  return std::filesystem::relative(Context::global.file_path, Context::global.input_folder).string();
}

string Generator::compiled_header_path() const
{
  return replace_extension(relative_filepath(), "hpp");
}

string Generator::compiled_source_path() const
{
  return replace_extension(relative_filepath(), "cpp");
}
